Update documentation for 'ml deployment' page to reflect new approaches to deployments (endpoints). (#1138)
MicrosoftDocs/cloud-adoption-framework: docs/innovate/best-practices/ml-deployment-inference.md

ms.custom: think-tank
---

# Machine learning (ML) inference during deployment

> Deploying machine learning models in production can be a complex process, but it's crucial for organizations that want to use artificial intelligence (AI) to enhance their operations. In this article, we'll explore some best practices for deploying ML models in production environments using Azure Machine Learning.

Reviewer: ```suggestion
Deploying machine learning models in production can be a complex process, but it's crucial for organizations that want to use artificial intelligence (AI) to enhance their operations. In this article, you'll explore some best practices for deploying ML models in production environments using Azure Machine Learning.
```

> ## Best Practices

Reviewer: We generally like to avoid this term, as it isn't meaningful to most.  Can you describe these considerations another way please?

> 1. **Choose the Right Deployment Method:**

Reviewer: Numbers should only be used for sequential steps, and not just "lists of things" that do not have an order.  This list doesn't look like it has an order.  Please change to bullets, or be much more clear about the ordering here.

> 1. **Choose the Right Deployment Method:**

Reviewer: ```suggestion
1. **Choose the right deployment method:**
```

> One of the first decisions you'll need to make is how to deploy your ML model. There are two main deployment methods to choose from; real-time, and batch. Each method has its own advantages and disadvantages, so it's important to choose the one that best suits your organization's needs.

Reviewer: ```suggestion
One of the first decisions you'll need to make is how to deploy your ML model. There are two main deployment methods to choose from:

   - real-time
   - batch

Each method has its own advantages and disadvantages, so it's important to choose the one that best suits your organization's needs.
```

>     - Real-time (or online) inference involves processing input data as it is received, often with a very low latency requirement. This is important for applications that require immediate responses, such as fraud detection, speech recognition, or recommendation systems. Real-time inference can be more complex and expensive to implement than batch inference because it requires a faster and more reliable infrastructure to process incoming requests and respond quickly.  Real-time inference usually requires the underlying compute to be running continuously to be able to respond to requests. 

Reviewer: ```suggestion
    - **Real-time** (or online) inference involves processing input data as it is received, often with a very low latency requirement. This is important for applications that require immediate responses, such as fraud detection, speech recognition, or recommendation systems. Real-time inference can be more complex and expensive to implement than batch inference because it requires a faster and more reliable infrastructure to process incoming requests and respond quickly.  Real-time inference usually requires the underlying compute to be running continuously to be able to respond to requests.
```

>     - Batch (or offline) processing involves processing a large batch of input data all at once, rather than processing each input data point individually in real time. Batch inference is well-suited for scenarios where a large volume of data needs to be processed efficiently and where the response time is not critical. For example, batch inference can be used for processing a large dataset of images, where predictions are made on all the images in the dataset at once. Batch inference is often less expensive and more efficient than real-time inference. Batch inference usually requires the underlying compute to be running only for the duration of the batch job.

Reviewer: ```suggestion
    - **Batch** (or offline) processing involves processing a large batch of input data all at once, rather than processing each input data point individually in real time. Batch inference is well-suited for scenarios where a large volume of data needs to be processed efficiently and where the response time is not critical. For example, batch inference can be used for processing a large dataset of images, where predictions are made on all the images in the dataset at once. Batch inference is often less expensive and more efficient than real-time inference. Batch inference usually requires the underlying compute to be running only for the duration of the batch job.
```

5. **Have a Plan for Updates:**
ML models are not static, and they will need to be updated over time as new data or new algorithms become available. It's important to have a plan in place for updating your model, including a process for testing and validating the updated model before deploying it in production.  Blue/green deployment is a common strategy for updating ML models in production. This involves deploying the updated model to a new environment, testing it, and then switching over to the new model once it's been validated. This ensures that your customers are not impacted by any issues with the updated model.  See [blue/green deployments using AzureML online endpoints](/azure/machine-learning/concept-endpoints#native-bluegreen-deployment)

Deploying machine learning models in production can be challenging, but following these best practices can help ensure a smooth and successful deployment. By choosing the right deployment method, ensuring consistency, monitoring performance, implementing security measures, and having a plan for updates, your organization can use AI to improve its operations and achieve its goals. 

> ## [Deployment Methods](#deployment-methods)

Reviewer: ```suggestion
## Deployment methods
```

- **Security requirements:** Use authentication and authorization to control access to the online endpoint for enhanced security.  
  - An online endpoint with ingress protection will only accept scoring requests from hosts inside a virtual network but not from the public internet. An online endpoint that is created in a private-link enabled workspace will have ingress protection.  See [Use network isolation for managed online endpoints](/azure/machine-learning/how-to-secure-online-endpoint) for more information.
  - Use Azure Active Directory tokens for control-plane authentication.  For data plane operations, [key-based and token-based](/azure/machine-learning/how-to-authenticate-online-endpoint) approaches are supported with the latter preferred due to having expiration on tokens.  Ensure access to retrieve the key or token for an online endpoint is restricted by Azure role-based access controls (Azure RBAC).

>   - Use SSL encryption on the endpoint.  This enabled by default for Azure ML endpoint invocation.<br></br>

Reviewer: ```suggestion
  - Use SSL encryption on the endpoint.  This enabled by default for Azure ML endpoint invocation.
```

- **Real time Integration:** Integrating Real time inference with other Azure services can be simplified, given the availability of SDKs for different languages and the ability to invoke the endpoint using a REST API.  The online endpoint can be invoked as part of an application's code.

- **MLflow:** If possible, use [MLflow with Azure Machine Learning](/azure/machine-learning/concept-mlflow) during model development.  Azure Machine Learning supports no-code deployment of models created and logged with MLflow.  When you [deploy your MLflow model to an online endpoint](/azure/machine-learning/how-to-deploy-mlflow-models-online-endpoints), you don't need to indicate a scoring script or an environment. 

## Additional Considerations

> ### ONNX
> Using the [Open Neural Network Exchange](https://onnx.ai/) (ONNX) can help optimise the inference of your machine learning models.  Optimising models for inference can be a challenge to tune to fully utilise hardware capabilities, particularly when using different kinds of platforms (cloud/edge, CPU/GPU, etc).  You can either train a new model, or convert an existing model from another format to ONNX.

Reviewer: ```suggestion
### ONNX

Using the [Open Neural Network Exchange](https://onnx.ai/) (ONNX) can help optimize the inference of your machine learning models.  Optimizing models for inference can be a challenge to tune to fully utilize hardware capabilities, particularly when using different kinds of platforms (cloud/edge, CPU/GPU, etc.).  You can either train a new model or convert an existing model from another format to ONNX.
```

**Model Training:** Several experiments run parallel to each data partition in the model training stage. This is a typically iterative process where models from the experiments are evaluated to determine the best one.

**Many models:** The best models for each segment or category are registered in the model registry. Assign meaningful names to the models, which will make them more discoverable for inference. Use tagging where necessary to group the model into specific categories.

## Batch inference for many models

During batch inference for many models, predictions are typically scheduled, recurring, and they can handle large volumes of data running at the same time. Unlike in a single-model scenario, many models inference at the same time, and it's important to select the correct ones. The following diagram shows the reference pattern for many-models batch inference:
During batch inference for many models, predictions are typically scheduled, recurring, and they can handle large volumes of data running at the same time. Unlike in a single-model scenario, many models inference at the same time, and it's important to select the correct ones. 

> [![Diagram of the reference pattern for many-models batch inference.](./media/many-models-batch-inference.png)](./media/many-models-batch-inference.png#lightbox)

Reviewer: If this image is no longer being used, it needs to be removed as part of the PR

Many models for batch inference can by achieved by using multiple deployments against a single Managed Endpoint.  Batch inference for specific models is achieved by invoking the deployment name during the REST or SDK call.  For an example of how to deploy multiple models to a single batch endpoint, see [Deploy multiple models to one deployment](https://github.com/Azure/azureml-examples/blob/main/sdk/python/endpoints/batch/deploy-models/mnist-classifier/mnist-batch.ipynb)

## Real-time inference for many models

Real-time many-models inference requires low latency and on-demand requests, typically via a REST endpoint. This is useful when external applications or services require a standard interface to interact with the model, typically via a REST interface with a JSON payload.

> [![Diagram of many-models real-time inference.](./media/many-models-real-time-inference.png)](./media/many-models-real-time-inference.png#lightbox)

Reviewer: If this image is no longer being used, it needs to be removed as part of the PR
